src/rust/runtime/memory/buffer/demibuffer.rs

```diff
@@ -0,0 +1,824 @@
+            let metadata: &mut MetaData = unsafe { temp.as_mut() };
+
+            // Point buf_addr at the newly allocated data space (if any).
+            if capacity == 0 {
+                // No direct data, so don't point buf_addr at anything.
+                metadata.buf_addr = null_mut();
+            } else {
+                // The direct data immediately follows the MetaData struct.
+                let address: *mut u8 = temp.cast::<u8>().as_ptr();
+                // Safety: The call to offset is safe, as the provided offset is known to be within the allocation.
+                metadata.buf_addr = unsafe { address.offset(size_of::<MetaData>() as isize) };
+            }
+
+            // Set field values as appropriate.
+            metadata.data_off = 0;
+            metadata.refcnt = 1;
+            metadata.nb_segs = 1;
+            metadata.ol_flags = 0;
+            metadata.pkt_len = capacity as u32;
+            metadata.data_len = capacity;
+            metadata.buf_len = capacity;
+            metadata.next = None;
+        }
+
+        // Embed the buffer type into the lower bits of the pointer.
+        let tagged: NonNull<MetaData> = temp.with_addr(temp.addr() | Tag::Heap);
+
+        // Return the new DemiBuffer.
+        DemiBuffer {
+            ptr: tagged,
+            _phantom: PhantomData,
+        }
+    }
+
+    // Create a new Heap-allocated DemiBuffer from a slice.
+    // Note: This is implemented as stand-alone function instead of a conversion (From) Trait in order to be able to
+    // handle the error case where the given slice is larger than a single DemiBuffer can hold.  ToDo: Review this?
+    pub fn from_slice(slice: &[u8]) -> Result<Self, Fail> {
+        // Check size of the slice to ensure a single DemiBuffer can hold it.
+        if slice.len() > u16::MAX as usize {
+            return Err(Fail::new(libc::EINVAL, "slice is larger than a DemiBuffer can hold"));
+        }
+        let amount: u16 = slice.len() as u16;
+
+        // Allocate some memory off the heap.
+        let mut temp: NonNull<MetaData> = allocate_metadata_data(amount);
+
+        // Initialize the MetaData.
+        {
+            // Safety: This is safe, as temp is aligned, dereferenceable, and metadata isn't aliased in this block.
+            let metadata: &mut MetaData = unsafe { temp.as_mut() };
+
+            // Point buf_addr at the newly allocated data space (if any).
+            if amount == 0 {
+                // No direct data, so don't point buf_addr at anything.
+                metadata.buf_addr = null_mut();
+            } else {
+                // The direct data immediately follows the MetaData struct.
+                let address: *mut u8 = temp.cast::<u8>().as_ptr();
+                // Safety: The call to offset is safe, as the provided offset is known to be within the allocation.
+                metadata.buf_addr = unsafe { address.offset(size_of::<MetaData>() as isize) };
```

Comment:
Why?  The [source code for `add`](https://doc.rust-lang.org/src/core/ptr/const_ptr.rs.html#871-873) literally just calls `offset` under the covers.  Sure, the compiler will likely optimize away the extra call, but why make it?

Reply:
Because you don't need the `isize` casting there.

Reply:
> * You are adding in the end of the day, so it makes code easier to follow.

I'd argue that the word "offset" is clearer in its intention (that you're staying inside the allocated region -- part of the Safety rules), so I strongly prefer to leave as is.
> * You don't need the casting there, which may hide potential address overflow computation.

The source code for `add` does the very same cast, which I'm surprised about actually, as they're hiding this conversion when it is better to leave it exposed to the programmer.  The [manual page for `add`](https://doc.rust-lang.org/std/primitive.pointer.html#method.add-1) has a Safety section (identical to `offset`'s) making it clear that you really have to think about this before calling the function (which is why it is marked unsafe in the first place).  It also explicitly states that `add` is just a convenience wrapper for `.offset(count as isize)`.  In any event, my exposed cast is clearer than hiding it in a call to `add`, which negates your point.

This discussion has made it pretty clear to me we should use `offset` and not `add`.

Reply:
> This discussion has made it pretty clear to me we should use offset and not add.

I disagree.

These two arguments go beyond coding style:

- With `add` you don't need to add a cast in our code, regardless how it is implemented on theirs.
- Casts may introduce bugs silently.

Reply:
> These two arguments go beyond coding style (ie. you you find it or not `offset` expresses better what is going on):
> 
> * With `add` you don't need to add a cast in our code, regardless how it is implemented on theirs.

While true, this is actually an argument for `offset`.  Since `add` is explicitly documented as being a shorthand for `.offset(count as isize)`, what you're saying is that using `add` hides the cast from the programmer (they have to read the documentation to figure out what it is doing).

> * Casts may introduce bugs silently.

Also true.  But again, this is argument for leaving them visible to the programmer, not hiding them in a wrapper call.

It's really clear to me that we should stick with `offset`.  And as a result of this discussion, I'm beginning to think we might want to consider replacing any use of `add` elsewhere in our codebase with `offset` as well.

Reply:
I cannot sign off in favor for casting.

Casting should be avoided, always. If we have an API that does not require casting, we should use it.

Furthermore, good code is the one that does not require one to opening up the API documentation.

Reply:
In the existing codebase (dev branch), we use "add" more often than "offset":

![image](https://user-images.githubusercontent.com/420175/195946791-c122f954-dbb7-4b1b-8647-68647ef0b904.png)

While both methods are not very different and not difficult to understand, I favor simplicity and consistency across the codebase. We should try to follow just 1 chain of thought so that it's the least surprising way of doing things in code. If we have consistency, we can easily search for this pattern and change if we ever decide to go the other way.

Also, add() does not allow negative offsets (not sure if we'd ever need that though).

Reply:
First off, we cast all over the place.  We're expected to know what we're doing when we are casting.  I disagree in general that we should always use a more dangerous API that doesn't require a cast instead of a safer API that does.  Casting is not universally bad; one just needs to know what you're doing when you do it.  This file alone has 50+ direct or indirect casts in it.

Second, the cast involved in this specific case is guaranteed to be safe.  It is casting the result of the `size_of` operator (a `usize`) to an `isize`.  For this to cast to be unsafe, the `size_of` the thing in question (here a `MetaData` struct) would need to exceed 2^31 bytes.  We don't have any structs even close to that, and this one is known to be only 128 bytes.

These two points alone are enough for me to conclude that there is nothing wrong with using `offset` instead of `add`.  For reasons previously stated, I would prefer to use `offset` over `add` even without the below.  But since you mention it...

> You are having these conclusions because you are reading the library implementation, which may change. We should use it as an opaque box.

Third, no, I'm aware of the cast in `add`'s implementation because I'm reading the library _documentation_ which explicitly states that `add` is just a wrapper around `offset` that makes this cast.  It is officially _defined_ as behaving that way.  The opaque box you speak of is around `offset`, not `add`.

> Furthermore, good code is the one that does not require one to opening up the API documentation.

I seriously hope you aren't writing any code using APIs for which you've never read the documentation.  If you're referring to readability of other's code, I would consider this as yet another argument for using `offset` over `add` in this case, as the cast is visible to the reader, and they don't have to open up the API documentation to know about it.

Reply:
My last reply was to Pedro's comment (GitHub doesn't make this easy to follow).

To Anand, I don't think number of lines of use of an API in the existing codebase really matters here.  I also don't think we need to necessarily choose between one or the other.  But if you're concerned about that, then for simplicity/consistency, I think using offset is both more simple and consistent once you realize that both `add` and `sub` are defined (and documented!) as simple wrappers around `offset`.  So you can use `offset` in place of both `add` and `sub` if you want the simplicity/consistency of a single API.

And yes, I have code that uses negative offsets, so we need to either use `offset` or `sub` for that.

Reply:
I think this is fine. I'm inclined to use the base function, rather than the convenience function, so that reading the code doesn't require following an indirection in the API documentation. While we can treat the API as a black box, the documentation explicitly points to add being a convenience function (and I'm inclined to not use convenience functions but that's just a personal preference that I don't want to impose). My recommendation is that we resolve this PR and make some "best practice" decisions separately and then open a separate housekeeping PR to clean up everything based on the decision.